ffi/derive/src/convert.rs

```diff
@@ -1,210 +1,236 @@
 use core::str::FromStr as _;
 
-use proc_macro2::TokenStream as TokenStream2;
+use proc_macro2::TokenStream;
 use proc_macro_error::{abort, OptionExt};
 use quote::quote;
-use syn::{parse_quote, DeriveInput, Generics, Ident};
+use syn::{parse_quote, Data, DataEnum, DeriveInput, Generics, Ident, Type};
 
-use crate::{find_attr, is_opaque, is_repr_attr};
+use crate::{find_attr, is_extern, is_opaque, is_repr_attr};
 
-pub fn derive_ffi_type(input: DeriveInput) -> TokenStream2 {
-    if !matches!(input.vis, syn::Visibility::Public(_)) {
-        abort!(input.vis, "Only public items are supported");
-    }
+pub fn derive_ffi_type(mut input: DeriveInput) -> TokenStream {
+    let local = is_local(&input.attrs, &input.data);
 
     let name = &input.ident;
+    if is_extern(&input.attrs) {
+        return derive_ffi_type_for_extern_item(name, &mut input.generics);
+    }
+    if is_opaque(&input) {
+        return derive_ffi_type_for_opaque_item(name, &mut input.generics);
+    }
     if is_transparent(&input) {
-        return derive_ffi_type_for_transparent_item(&input);
+        return derive_ffi_type_for_transparent_item(&mut input);
     }
 
     match input.data {
-        syn::Data::Enum(item) => {
+        Data::Enum(item) => {
             let repr = find_attr(&input.attrs, "repr");
 
+            if item.variants.is_empty() {
+                abort!(name, "uninhabited enum's cannot be instantiated");
+            }
+
             if is_fieldless_enum(&item) {
                 derive_ffi_type_for_fieldless_enum(&input.ident, &item, &repr)
             } else {
-                derive_ffi_type_for_data_carrying_enum(
-                    &input.ident,
-                    &input.attrs,
-                    input.generics,
-                    &item,
-                )
+                derive_ffi_type_for_data_carrying_enum(&input.ident, input.generics, &item, local)
             }
         }
-        syn::Data::Struct(_) => {
-            if is_opaque(&input) {
-                derive_ffi_type_for_opaque_struct(name, input.generics)
-            } else if is_opaque_wrapper(&input.attrs) {
-                derive_ffi_type_for_extern_struct(name, input.generics)
-            } else {
-                derive_ffi_type_for_repr_c_struct(name)
+        Data::Struct(item) => {
+            if local {
+                abort!(name, "if applicable, attach `#[ffi_type(non_local)`")
```

Comment:
This error give 0 information on what went wrong, so developer want be able to decide attach attribute or not.

Reply:
I think we need to view it in context. It gives you a hint, that you're missing `#[ffi_type(non_local)]`, but I agree that it should give the reader a bit more context.